### ISSUE TITLE
[bug] static runtime is ignored for msvc compiler

### ISSUE THREAD

**Issue by Sil3ntStorm** [conan-io/conan#10299]
### Environment Details (include every applicable attribute)
  * Operating System+version: Windows 10
  * Compiler+version: MSVC 19.31.30919.0
  * Conan version: 1.44.0
  * Python version: 3.9.7

### Steps to reproduce (Include if Applicable)
Attempt to build a package (those I tested used cmake under the hood) with `settings.compiler.runtime=static` using the latest msvc compiler. When using `settings.compiler.runtime_type=Debug` it should build with `/MTd`, however it actually uses `/MDd`. When using `settings.compiler.runtime_type=Release` it should build with `/MT`, however it actually uses `/MD`.

The `/MDd` or `/MD` flavor should only be used when the `settings.compiler.runtime` configuration is set to dynamic, otherwise it should use `/MT` or `/MTd`

It appears to be impossible (or I have not yet found a way) to convince conan to use the desired runtime.

**Comment by memsharded** [comment]
Hi @Sil3ntStorm 

The settings are applied in recipes by the build helpers, so it depends on the build helper used: is it ``CMakeToolchain``?

Also, the legacy integrations might not fully support this new experimental ``msvc`` compiler, and you need to use only the new integrations (``from conan.tools.cmake``, ``from conan.tools.microsoft``) to get full support for ``msvc``. Please check https://docs.conan.io/en/latest/reference/conanfile/tools.html, and report about the used integrations. I recommend to start from a ``conan new hello/0.1 --template=cmake_lib`` and report from there.

**Comment by Sil3ntStorm** [comment]
The recipe is using `from conans import CMake`, a quick test to change that leads to all sorts of weird (usage) errors, so I will need to look into those.
By default conan creates a profile that detects the compiler as msvc however, will see if I can convince it to do the right thing when forcing it back to "Visual Studio"

**Comment by memsharded** [comment]
> The recipe is using from conans import CMake, a quick test to change that leads to all sorts of weird (usage) errors, so I will need to look into those.

Yes, that ``from conans import CMake`` is one of the previous build integrations, you need to use ``Visual Studio`` compiler not ``msvc`` with it. It is possible the autodetection for latest VS is returning msvc, because it is already moving forward, in the same way the build integrations are moving forward to the new ones.

**Comment by Sil3ntStorm** [comment]
When setting the compiler in the profile as "Visual Studio" I do however get an error (without any error message) that building the recipe failed.
Manually executing the command it attempts to execute reveals an error between the compiler version as detected and as specified.
```
CMake Error at conanbuildinfo.cmake:511 (message):
  Detected a mismatch for the compiler version between your conan profile
  settings and CMake:

  Compiler version specified in your conan profile: 17

  Compiler version detected in CMake: 19.31
```
However 17 is the highest compiler version conan knows of.

This can be worked around by adding `cmake.definitions['CONAN_DISABLE_CHECK_COMPILER'] = True` in the recipe, this however is not present in official recipes I use, which will consequently fail to build.

Is there any way to have this specified in the profile / globally?

PS: This is turning out to be a major headache :(

**Comment by memsharded** [comment]
Something seems wrong in the conanbuildinfo.cmake compiler detection. Can you please post the full output of the command, including the initial profile that is printed?

Also the commands that you are using, are you using ``conan build`` by any chance? It is possible that the result of the previous ``conan install`` is being there cached, make sure to clean temporary ``conaninfo.txt`` and issue ``conan install`` again.

> PS: This is turning out to be a major headache :(

Yes, you are trying to upgrade to the latest VS 2022, and most of our efforts are now focused on the new integrations and releasing Conan 2.0, so the legacy integrations are not prepared to handle the very latest tools.

**Comment by Sil3ntStorm** [comment]
I am trying to run conan install --build=missing, so yes it is attempting to build one dependency as that does not seem to have a prebuilt version for Visual Studio version 17.

```
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Debug
compiler=Visual Studio
compiler.cppstd=17
compiler.runtime=MTd
compiler.version=17
os=Windows
os_build=Windows
[options]
[build_requires]
[env]

zlib/1.2.11: Configuring sources in <redacted>\.conan\data\zlib\1.2.11\_\_\source
Downloading zlib-1.2.11.tar.gz completed [593.46k]

zlib/1.2.11: Copying sources to build folder
zlib/1.2.11: Building your package in <redacted>\.conan\data\zlib\1.2.11\_\_\build\3682397a88110bb423c0b4c7c75a818b7fdf9bd9
zlib/1.2.11: Generator cmake created conanbuildinfo.cmake
zlib/1.2.11: Aggregating env generators
zlib/1.2.11: Calling build()
-- Selecting Windows SDK version 10.0.20348.0 to target Windows 10.0.19042.
-- The C compiler identification is MSVC 19.31.30919.0
-- The CXX compiler identification is MSVC 19.31.30919.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Preview/VC/Tools/MSVC/14.31.30919/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Preview/VC/Tools/MSVC/14.31.30919/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: called by CMake conan helper
-- Conan: called inside local cache
-- Conan: Adjusting output directories
-- Conan: Using cmake global configuration
-- Conan: Adjusting default RPATHs Conan policies
-- Conan: Adjusting language standard
CMake Error at conanbuildinfo.cmake:402 (message):
  Detected a mismatch for the compiler version between your conan profile
  settings and CMake:

  Compiler version specified in your conan profile: 17

  Compiler version detected in CMake: 19.31

  Please check your conan profile settings (conan profile show
  [default|your_profile_name])

  P.S.  You may set CONAN_DISABLE_CHECK_COMPILER CMake variable in order to
  disable this check.
Call Stack (most recent call first):
  conanbuildinfo.cmake:470 (conan_error_compiler_version)
  conanbuildinfo.cmake:591 (check_compiler_version)
  conanbuildinfo.cmake:141 (conan_check_compiler)
  CMakeLists.txt:5 (conan_basic_setup)


-- Configuring incomplete, errors occurred!
See also "<redacted>/.conan/data/zlib/1.2.11/_/_/build/3682397a88110bb423c0b4c7c75a818b7fdf9bd9/build_subfolder/CMakeFiles/CMakeOutput.log".
zlib/1.2.11:
zlib/1.2.11: ERROR: Package '3682397a88110bb423c0b4c7c75a818b7fdf9bd9' build failed
```

**Comment by memsharded** [comment]
Ok, thanks for the feedback.

Now I see the bug. The legacy ``cmake`` generator has been only updated to 19.30, but the latest 19.31 is not covered:
```cmake
# 1930 = VS 17.0 (v143 toolset)
                    (CONAN_COMPILER_VERSION STREQUAL "17" AND NOT MSVC_VERSION EQUAL 1930) OR
```

As commented above, these build system integrations are no longer under active development, and only the new ones are, this is kind of expected. Lets try to do a fix for next 1.45 release.

**Comment by Sil3ntStorm** [comment]
Even after changing those lines to make it so that it won't fail the compiler detection, it will not detect a C compiler either.

```
-- Conan: called by CMake conan helper
-- Conan: called inside local cache
-- Conan: Adjusting output directories
-- Conan: Using cmake global configuration
-- Conan: Adjusting default RPATHs Conan policies
-- Conan: Adjusting language standard
-- Conan setting CPP STANDARD: 17 WITH EXTENSIONS OFF
CMake Error at conanbuildinfo.cmake:648 (message):
  This project seems to be plain C, but no compiler defined
Call Stack (most recent call first):
  conanbuildinfo.cmake:250 (conan_check_compiler)
  CMakeLists.txt:3 (conan_basic_setup)


-- Configuring incomplete, errors occurred!
```

**Comment by memsharded** [comment]
I am submitting https://github.com/conan-io/conan/pull/10361 to fix the cmake check for vs 2022.

The other problem is related to the usage of CXX and C compilers, by default the ``cmake`` generator assumes using a CXX compiler. If it is not the case, the check might fail. Please keep using the ``CONAN_DISABLE_CHECK_COMPILER`` to disable the compiler check, this won't be further fixed, as commented, this generator is already removed in 2.0, it will be no longer maintained, and it will be necessary to use ``CMakeToolchain``, ``CMakeDeps``.

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/10361 merged, will be in 1.45.

As a summary note regarding the issue title: to use modern ``msvc`` compiler definition, the new integrations are necessary (``from conan.tools.cmake``).

**Comment by Sil3ntStorm** [comment]
Thanks for the fix. This will make legacy cmake generator properly detect VS2022 as an available compiler?
I assume that the static runtime issue being ignored is because the legacy cmake generator does not know of these new flags, which is why it won't do static runtime?
If that's the case, I guess the workaround would be to force the profile to use the "Visual Studio" compiler with the old runtime settings for that compiler.

What are the schedule estimates on Conan 2.0 / conan center recipes using the new integrations?
Also out of curiosity what determines which kind of builds are available on conan center?
